crates/factors/src/factor.rs

```diff
@@ -0,0 +1,152 @@
+use std::any::Any;
+
+use anyhow::Context;
+
+use crate::{App, FactorInstancePreparer, Linker, ModuleLinker, RuntimeConfig, SpinFactors};
+
+pub trait Factor: Any + Sized {
+    /// Per-app configuration for this factor.
+    ///
+    /// See [`Factor::configure_app`].
+    type AppConfig: Default;
+
+    /// The [`FactorInstancePreparer`] for this factor.
+    type InstancePreparer: FactorInstancePreparer<Self>;
```

Comment:
Sketching out what that could look like:

```rust
trait Factor {
  ...

  type InstancePreparer: Default;
  fn create_preparer(&self, ...) -> Result<Self::InstancePreparer> {
    Ok(Default::default())
  }

  // Either:
  type InstanceState;
  fn prepare(&self, preparer: Self::InstancePreparer) -> Result<Self::InstanceState>;
  // Or:
  type InstanceState: Default;
  fn prepare(&self, preparer: Self::InstancePreparer) -> Result<Self::InstanceState> {
    Ok(Default::default())
  }
```

The annoying thing here is `prepare`; you either need to have no default impl _or_ you need to make `InstanceState: Default`, both of which may be annoying for different factors.

That said, I was already on the fence about how to do this so given your feedback I would probably now lean toward one of these annoying ways rather than the existing annoying way. :slightly_smiling_face:

Reply:
Done (first option)